src/embedded_data.cc

```diff
@@ -0,0 +1,135 @@
+// Copyright (C) 2022 Toitware ApS.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; version
+// 2.1 only.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// The license can be found in the file `LICENSE` in the top level
+// directory of this repository.
+
+#include "embedded_data.h"
+#include "entropy_mixer.h"
+#include "uuid.h"
+
+namespace toit {
+
+const EmbeddedDataExtension* EmbeddedDataExtension::cast(const void* pointer) {
+  const uint32* header = reinterpret_cast<const uint32*>(pointer);
+  if (!header) return null;
+  if (header[HEADER_INDEX_MARKER] != HEADER_MARKER) return null;
+  uint32 checksum = 0;
+  for (int i = 0; i < HEADER_WORDS; i++) checksum ^= header[i];
+  if (checksum != HEADER_CHECKSUM) return null;
+  return reinterpret_cast<const EmbeddedDataExtension*>(header);
+}
+
+int EmbeddedDataExtension::images() const {
+  const uint32* header = reinterpret_cast<const uint32*>(this);
+  return header[HEADER_INDEX_IMAGE_COUNT];
+}
+
+EmbeddedImage EmbeddedDataExtension::image(int n) const {
+  ASSERT(n >= 0 && n < images());
+  const uint32* header = reinterpret_cast<const uint32*>(this);
+  const uword* table = reinterpret_cast<const uword*>(&header[HEADER_WORDS]);
+  return EmbeddedImage{
+    .program = reinterpret_cast<const Program*>(table[n * 2]),
+    .size    = table[n * 2 + 1]
+  };
+}
+
+List<uint8> EmbeddedDataExtension::config() const {
+  const uint32* header = reinterpret_cast<const uint32*>(this);
+  uint32 used = header[HEADER_INDEX_USED];
+  uint32 free = header[HEADER_INDEX_FREE];
+  if (free < sizeof(uint32)) return List<uint8>();
+  uword address = reinterpret_cast<uword>(header) + used;
+  uint32 size = *reinterpret_cast<uint32*>(address);
+  uint8* data = reinterpret_cast<uint8*>(address + sizeof(uint32));
+  return List<uint8>(data, Utils::min(size, free));
```

Comment:
The two values in the `min` are not comparable.

Reply:
Done.